cirq-core/cirq/ops/raw_types_test.py

```diff
@@ -739,3 +739,184 @@ def qubits(self):
+    assert c.on_each([a, b]) == [c(a), c(b)]
+    assert c.on_each([b, a]) == [c(b), c(a)]
+    assert c.on_each([a, [b, a], b]) == [c(a), c(b), c(a), c(b)]
+
+    with pytest.raises(ValueError):
+        c.on_each('abcd')
+    with pytest.raises(ValueError):
+        c.on_each(['abcd'])
+    with pytest.raises(ValueError):
+        c.on_each([a, 'abcd'])
+
+    def iterator(qubits):
+        for i in range(len(qubits)):
+            yield qubits[i]
+
+    qubit_iterator = iterator([a, b, a, b])
+    assert isinstance(qubit_iterator, Iterator)
+    assert c.on_each(qubit_iterator) == [c(a), c(b), c(a), c(b)]
+
+
+def test_on_each_two_qubits():
+    class CustomGate(cirq.ops.gate_features.SupportsOnEachGate, cirq.TwoQubitGate):
+        pass
+
+    a = cirq.NamedQubit('a')
+    b = cirq.NamedQubit('b')
+    g = CustomGate()
+
+    assert g.on_each([]) == []
+    assert g.on_each([(a, b)]) == [g(a, b)]
+    assert g.on_each([[a, b]]) == [g(a, b)]
+    assert g.on_each([(b, a)]) == [g(b, a)]
+    assert g.on_each([(a, b), (b, a)]) == [g(a, b), g(b, a)]
+    assert g.on_each(zip([a, b], [b, a])) == [g(a, b), g(b, a)]
+    assert g.on_each() == []
+    assert g.on_each((b, a)) == [g(b, a)]
+    assert g.on_each((a, b), (a, b)) == [g(a, b), g(a, b)]
+    assert g.on_each(*zip([a, b], [b, a])) == [g(a, b), g(b, a)]
+    with pytest.raises(TypeError, match='object is not iterable'):
+        g.on_each(a, b)
+    with pytest.raises(ValueError, match='Inputs to multi-qubit gates must be Sequence'):
+        g.on_each([12])
+    with pytest.raises(ValueError, match='Inputs to multi-qubit gates must be Sequence'):
+        g.on_each([(a, b), 12])
+    with pytest.raises(ValueError, match='All values in sequence should be Qids'):
+        g.on_each([(a, b), [(a, b)]])
+    with pytest.raises(ValueError, match='Expected 2 qubits'):
+        g.on_each([()])
+    with pytest.raises(ValueError, match='Expected 2 qubits'):
+        g.on_each([(a,)])
+    with pytest.raises(ValueError, match='Expected 2 qubits'):
+        g.on_each([(a, b, a)])
+    with pytest.raises(ValueError, match='Expected 2 qubits'):
+        g.on_each(zip([a, a]))
+    with pytest.raises(ValueError, match='Expected 2 qubits'):
+        g.on_each(zip([a, a], [b, b], [a, a]))
+    with pytest.raises(ValueError, match='All values in sequence should be Qids'):
+        g.on_each('ab')
+    with pytest.raises(ValueError, match='All values in sequence should be Qids'):
+        g.on_each(('ab',))
+    with pytest.raises(ValueError, match='All values in sequence should be Qids'):
+        g.on_each([('ab',)])
+    with pytest.raises(ValueError, match='All values in sequence should be Qids'):
+        g.on_each([(a, 'ab')])
+    with pytest.raises(ValueError, match='All values in sequence should be Qids'):
+        g.on_each([(a, 'b')])
+
+    def iterator(qubits):
+        for i in range(len(qubits)):
+            yield qubits[i]
+
+    qubit_iterator = iterator([[a, b], [a, b]])
```

Comment:
here, too, can use generator expression
```suggestion
    qubit_iterator = (qs for qs in [[a, b], [a, b]])
```